# Dataset Audit

This notebook is used to understand the original IMDb dataset before any cleaning.

Goals:

- Understand the dataset structure
- Inspect data types
- Explore missing values
- Check duplicates
- Identify potential data quality issues

1. Load Dataset
2. Dataset Shape
3. Dataset Preview
4. Missing Values
5. Duplicate Rows
6. Data Type Inspection
7. Data Quality Checks
    *     Numeric Columns
    *     Date Columns
    *     Boolean Columns
    *     JSON Columns
8. Initial Observations 

In [1]:
import pandas as pd
import numpy as np

## Load Dataset

In [2]:
original_data = pd.read_csv('data/raw.csv', low_memory=False)
data = original_data.copy()

## Dataset Shape

In [3]:
print(f"Rows: {data.shape[0]}")
print(f"columns: {data.shape[1]}")
print("\ndata info")
data.info()

Rows: 45466
columns: 24

data info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date       

## Dataset Preview

In [4]:
data.head()

,adult,belongs_to_collection,budget,genres,homepage,id,imdb_id,original_language,original_title,overview,...,release_date,revenue,runtime,spoken_languages,status,tagline,title,video,vote_average,vote_count
0,False,"{'id': 10194, 'name': 'Toy Story Collection', ...",30000000,"[{'id': 16, 'name': 'Animation'}, {'id': 35, '...",http://toystory.disney.com/toy-story,862,tt0114709,en,Toy Story,"Led by Woody, Andy's toys live happily in his ...",...,1995-10-30,373554033.0,81.0,"[{'iso_639_1': 'en', 'name': 'English'}]",Released,NaN,Toy Story,False,7.7,5415.0
1,False,NaN,65000000,"[{'id': 12, 'name': 'Adventure'}, {'id': 14, '...",NaN,8844,tt0113497,en,Jumanji,When siblings Judy and Peter discover an encha...,...,1995-12-15,262797249.0,104.0,"[{'iso_639_1': 'en', 'name': 'English'}, {'iso...",Released,Roll the dice and unleash the excitement!,Jumanji,False,6.9,2413.0
2,False,"{'id': 119050, 'name': 'Grumpy Old Men Collect...",0,"[{'id': 10749, 'name': 'Romance'}, {'id': 35, ...",NaN,15602,tt0113228,en,Grumpier Old Men,A family wedding reignites the ancient feud be...,...,1995-12-22,0.0,101.0,"[{'iso_639_1': 'en', 'name': 'English'}]",Released,Still Yelling. Still Fighting. Still Ready for...,Grumpier Old Men,False,6.5,92.0
3,False,NaN,16000000,"[{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...",NaN,31357,tt0114885,en,Waiting to Exhale,"Cheated on, mistreated and stepped on, the wom...",...,1995-12-22,81452156.0,127.0,"[{'iso_639_1': 'en', 'name': 'English'}]",Released,Friends are the people who let you be yourself...,Waiting to Exhale,False,6.1,34.0
4,False,"{'id': 96871, 'name': 'Father of the Bride Col...",0,"[{'id': 35, 'name': 'Comedy'}]",NaN,11862,tt0113041,en,Father of the Bride Part II,Just when George Banks has recovered from his ...,...,1995-02-10,76578911.0,106.0,"[{'iso_639_1': 'en', 'name': 'English'}]",Released,Just When His World Is Back To Normal... He's ...,Father of the Bride Part II,False,5.7,173.0


In [5]:
data.sample(5)

,adult,belongs_to_collection,budget,genres,homepage,id,imdb_id,original_language,original_title,overview,...,release_date,revenue,runtime,spoken_languages,status,tagline,title,video,vote_average,vote_count
37487,False,"{'id': 380160, 'name': 'Die Kirche bleibt im D...",0,"[{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...",NaN,367596,tt5133288,de,Täterätää - Die Kirche bleibt im Dorf 2,NaN,...,2015-06-25,0.0,NaN,[],Released,NaN,Täterätää - Die Kirche bleibt im Dorf 2,False,4.5,4.0
26559,False,"{'id': 87096, 'name': 'Avatar Collection', 'po...",0,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...",http://www.avatarmovie.com/,76600,tt1630029,en,Avatar 2,A sequel to Avatar (2009).,...,2020-12-16,0.0,0.0,"[{'iso_639_1': 'en', 'name': 'English'}, {'iso...",In Production,NaN,Avatar 2,False,0.0,58.0
30798,False,NaN,0,"[{'id': 27, 'name': 'Horror'}]",NaN,30926,tt0117342,en,Pinocchio's Revenge,Defense attorney Jennifer Garrick acquires a P...,...,1996-10-07,0.0,96.0,"[{'iso_639_1': 'en', 'name': 'English'}, {'iso...",Released,Evil comes with strings attached.,Pinocchio's Revenge,False,4.3,13.0
16213,False,NaN,0,"[{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...",NaN,38340,tt0059269,en,A High Wind in Jamaica,"In 1870, a Jamaican colonial family sends its ...",...,1965-06-04,0.0,103.0,"[{'iso_639_1': 'en', 'name': 'English'}]",Released,NaN,A High Wind in Jamaica,False,5.5,9.0
22762,False,"{'id': 252338, 'name': 'The Four Collection', ...",0,"[{'id': 28, 'name': 'Action'}]",NaN,120143,tt2204340,zh,四大名捕,An undercover agent is assigned by his corrupt...,...,2012-07-12,0.0,118.0,"[{'iso_639_1': 'vi', 'name': 'Tiếng Việt'}]",Released,NaN,The Four,False,6.5,20.0


- The original_title column contains non-English titles
- The adult column is predominantly False
- some columns contain nested JSON-like strings (`belongs_to_collection`, `genres`, `spoken_languages`)

## Missing data

In [6]:
missing_data = pd.concat([
    data.isnull().sum(),
    round(data.isnull().mean() * 100, 3)], 
    axis=1, 
    keys=['count', 'percentage'])
missing_data.sort_values('percentage', ascending=False)

,count,percentage
belongs_to_collection,40972,90.116
homepage,37684,82.884
tagline,25054,55.105
overview,954,2.098
poster_path,386,0.849
runtime,263,0.578
status,87,0.191
release_date,87,0.191
imdb_id,17,0.037
original_language,11,0.024


## duplicate data

In [27]:
duplicate_rows = data.duplicated().sum()
print(f"Duplicate rows: {duplicate_rows}")

Duplicate rows: 17


## Data Type Inspection

In [ ]:
expected_dtypes = {
    # Numerical
    'budget': 'int64',
    'popularity': 'float64',
    'revenue': 'int64',
    'runtime': 'int64',
    'vote_average': 'float64',
    'vote_count': 'int64',
    
    # Categorical
    'original_language': 'object',
    'status': 'object',
    
    # Boolean
    'adult': 'bool',
    'video': 'bool',
    
    # String / Text
    'original_title': 'object',
    'title': 'object',
    'overview': 'object',
    'tagline': 'object',
    'homepage': 'object',
    'poster_path': 'object',
    'imdb_id': 'object',
    
    # JSON / Nested
    'belongs_to_collection': 'object',
    'genres': 'object',
    'production_companies': 'object',
    'production_countries': 'object',
    'spoken_languages': 'object',
    
    # Identifier
    'id': 'int64',
    
    # Date
    'release_date': 'datetime64[ns]'
}
actual_dtypes = data.dtypes.to_dict()

In [26]:
# DataFrame to compare
dtype_comparison = pd.DataFrame({
    'Column': data.columns,
    'Current_Dtype': data.dtypes.values,
    'Expected_Dtype': [expected_dtypes.get(col, 'Unknown') for col in data.columns],
    'Match?': [data.dtypes[col] == expected_dtypes.get(col, '') for col in data.columns]
})

dtype_comparison
# Display inconsistent columns
dtype_comparison[~dtype_comparison['Match?']]

,Column,Current_Dtype,Expected_Dtype,Match?
0,adult,object,bool,False
2,budget,object,int64,False
5,id,object,int64,False
10,popularity,object,float64,False
14,release_date,object,datetime64[ns],False
15,revenue,float64,int64,False
16,runtime,float64,int64,False
21,video,object,bool,False
23,vote_count,float64,int64,False


### Issues Found:
**Boolean Columns (object → bool)**

`adult`, `video`: Currently stored as object/text, need conversion to boolean

**Numeric Columns**

`budget`, `id`, popula`rity: Currently object, should be numeric `(`int64`/`float64`)`

`revenue`, `runtime`, `vote_count`: Currently `float64`, should be `int64`

**Date Column**

`release_date`: Currently object, should be `datetime64[ns]`

- Note: JSON-like columns (`belongs_to_collection`, `genres`, `production_companies`, `production_countries`, `spoken_languages`) need parsing and flattening.

## Data Quality Checks

In [50]:
# Boolean
print(data['video'].value_counts())
print(f"\n{data['adult'].value_counts()}")

video
False    45367
True        93
Name: count, dtype: int64

adult
False                                                                                                                             45454
True                                                                                                                                  9
 - Written by Ørnås                                                                                                                   1
 Rune Balot goes to a casino connected to the October corporation to try to wrap up her case once and for all.                        1
 Avalanche Sharks tells the story of a bikini contest that turns into a horrifying affair when it is hit by a shark avalanche.        1
Name: count, dtype: int64


In [151]:
# Categorical - Validate JSON columns
def analyze_non_dict_columns(df, columns_to_check):
    results = []
    for col in columns_to_check:
        non_empty = df[col][df[col].notna() & (df[col].astype(str).str.strip() != '')]
        invalid = non_empty[~non_empty.apply(lambda x: isinstance(x, list))]
        
        total = len(non_empty)
        count = len(invalid)
        results.append({
            'column': col,
            'total_non_empty': total,
            'invalid_count': count,
            'invalid_percent': round((count / total * 100), 2) if total > 0 else 0
        })
    return pd.DataFrame(results)

columns_to_check = ['belongs_to_collection', 'genres', 'production_companies', 
                    'production_countries', 'spoken_languages']
result_df = analyze_non_dict_columns(data, columns_to_check)
print(result_df)


                  column  total_non_empty  invalid_count  invalid_percent
0  belongs_to_collection             4494           4494            100.0
1                 genres            45466          45466            100.0
2   production_companies            45463          45463            100.0
3   production_countries            45463          45463            100.0
4       spoken_languages            45460          45460            100.0


In [150]:
# Categorical
def analyze_non_dict_columns(df, columns_to_check):
    results = []
    for col in columns_to_check:
        non_empty = df[col][df[col].notna() & (df[col].astype(str).str.strip() != '')]
        not_dict = non_empty[~non_empty.apply(lambda x: isinstance(x, str))]
        total = len(non_empty)
        count = len(not_dict)
        results.append({
            'column': col,
            'total_non_empty': total,
            'non_dict_count': count,
            'non_dict_percent': round((count / total * 100), 2) if total > 0 else 0
        })
    return pd.DataFrame(results)

columns_to_check = ['belongs_to_collection', 'genres', 'production_companies', 
                    'production_countries', 'spoken_languages']
result_df = analyze_non_dict_columns(data, columns_to_check)
print(result_df)

                  column  total_non_empty  non_dict_count  non_dict_percent
0  belongs_to_collection             4494               0               0.0
1                 genres            45466               0               0.0
2   production_companies            45463               0               0.0
3   production_countries            45463               0               0.0
4       spoken_languages            45460               0               0.0


In [105]:
# Date
total = data["release_date"].notna().sum()
valid_dates = pd.to_datetime(data["release_date"], errors='coerce').notna().sum()
invalid_count = total - valid_dates
print(f"invalid dates: {invalid_count}")

# Check invalid dates (future or incorrect month/day)
total_dates = pd.to_datetime(data["release_date"], errors='coerce')
invalid_mask = (total_dates > pd.Timestamp.now()) | (total_dates.dt.month > 12) | (total_dates.dt.day > 31)
print(f"Invalid dates (future or wrong month/day): {invalid_mask.sum()}")

invalid dates: 3
Invalid dates (future or wrong month/day): 0


In [137]:
# Numerical
rules = {
    'runtime': lambda x: (pd.to_numeric(x, errors='coerce') == 0),
    'budget': lambda x: pd.to_numeric(x, errors='coerce') == 0,
    'revenue': lambda x: pd.to_numeric(x, errors='coerce') == 0,
    'popularity': lambda x: pd.to_numeric(x, errors='coerce') == 0,
    'vote_average': lambda x: (pd.to_numeric(x, errors='coerce') == 0) | (pd.to_numeric(x, errors='coerce') > 10),
    'vote_count': lambda x: pd.to_numeric(x, errors='coerce') == 0,
    'id': lambda x: pd.to_numeric(x, errors='coerce') == 0,
}

# Create results DataFrame
results = []
for col, rule in rules.items():
    if col in data.columns:
        invalid_count = rule(data[col]).sum()
        total = data[col].notna().sum()
        results.append({
            'column': col,
            'invalid_count': invalid_count,
            'invalid_percent': round((invalid_count / total) * 100, 2)
        })

pd.DataFrame(results)


,column,invalid_count,invalid_percent
0,runtime,1558,3.45
1,budget,36573,80.44
2,revenue,38052,83.70
3,popularity,66,0.15
4,vote_average,2998,6.59
5,vote_count,2899,6.38
6,id,0,0.00


## Initial Observations

### Data Quality Summary

| Check | Result |
|-------|--------|
| Total Rows | 45,466 |
| Total Columns | 24 |
| Duplicate Rows | 17 |
| Invalid Dates | 3 |
| Future/Invalid Dates | 0 |
| JSON Invalid Values | 0 |
| Negative Values | 0 |

---

### Data Type Issues (9 Columns Need Conversion)

| Column | Current → Expected |
|--------|-------------------|
| `adult`, `video` | object → bool |
| `budget`, `id`, `popularity` | object → int64/float64 |
| `revenue`, `runtime`, `vote_count` | float64 → int64 |
| `release_date` | object → datetime64[ns] |

---

### Missing Data (High Priority)

| Column | Missing % | Action |
|--------|-----------|--------|
| `belongs_to_collection` | 90.1% | Consider dropping |
| `homepage` | 82.9% | Consider dropping |
| `tagline` | 55.1% | Consider dropping |
| `overview` | 2.1% | Impute or keep |
| `runtime` | 0.6% | Impute |
| Other columns | < 0.2% | Minor impact |

---

### Zero Value Analysis

| Column | Zero Count | Zero % |
|--------|------------|--------|
| `budget` | 36,573 | 80.4% |
| `revenue` | 38,052 | 83.7% |
| `runtime` | 1,558 | 3.4% |
| `vote_average` | 2,998 | 6.6% |
| `vote_count` | 2,899 | 6.4% |
| `popularity` | 66 | 0.2% |
| `id` | 0 | 0.0% |

---

### Key Findings

- **Data Contamination**: 

3 rows in `adult` column contain text/overview data (likely import shift)

- **JSON Columns**: 

All valid (`belongs_to_collection`, `genres`, `production_companies`, `production_countries`, `spoken_languages`) → ready for parsing

- **Titles**: 

`original_title` contains non-English titles (German, Chinese, etc.)

- **Adult Flag**: 

45,454 records are `False`, only 9 are `True`

- **High Zero Values**: 

Budget and revenue show >80% zero values → may indicate missing/unreported data

---

### TODO Actions

• Dataset contains both structured and JSON-like columns.

• release_date should be converted to datetime.

• JSON-like columns require parsing before analysis.

• budget and revenue contain many zero values that may represent missing information.

• runtime contains invalid values and requires cleaning.

• Several text columns have a high percentage of missing values.

• The dataset is suitable for further cleaning and exploratory analysis.